rfcs/shared/build-system/03-packaging-for-npm.md

```diff
@@ -164,17 +172,89 @@ This approach significantly improves our current situation:
 
-### 2. shipping only type definition rollup
+### 2. shipping type declarations (TypeScript)
 
 > **NOTE:** to make this RFC more focused this will be covered in followup PR's if needed
 
+#### Problem statement
+
+To be able to provide TypeScript support to our consumers we generate declaration files during `tsc` invocation.
+
+These declaration files are:
+
+1. generated side by side with implementation files
+1. generated N times for N supported module loaders (for v9 N equals to 2 - esm and commonjs)
+
+Aforementioned approach introduces 2 problems.
+
+**Problem 1: generated side by side with implementation files**
+
+- exposure of all package internals and structure and allows consumers to use **private APIs**
+- exposure of all TypeScript constructs that might violate consumers TypeScript version restrictions
+  - eg. internally in the package we use Template literal types that are not exposed to public API. If users depend on TS version that doesn't support this language feature they will get TS errors thus won't be able to ship code.
+
+**Problem 2: generated N times for N supported module loaders**
+
+- generated TypeScript declaration files don't emit different output based on different `module` field, thus this generation is redundant and provides no benefit.
+
+_Example:_
+
+_`module: CommonJs`_
+
+```ts
+import type { FunctionComponent } from 'react';
+import { TextWrapperProps } from '../wrapper';
+export declare const bodyClassName = 'fui-Body';
+export declare const Body: FunctionComponent<TextWrapperProps>;
+```
+
+_`module: ESM`_
+
+```ts
+import type { FunctionComponent } from 'react';
+import { TextWrapperProps } from '../wrapper';
+export declare const bodyClassName = 'fui-Body';
+export declare const Body: FunctionComponent<TextWrapperProps>;
+```
+
+#### Detailed Design or Proposal
+
+**Problem 1 solution:**
+
+- ship only [rolluped declaration files `.d.ts`](https://medium.com/@martin_hotell/typescript-library-tips-rollup-your-types-995153cc81c7) for all public API entrances. We can leverage already existing setup via api-extractor.
+
+Execution flow will look like following:
+
+![rolluping execution flow](https://miro.medium.com/max/1400/1*NGJGYhg9YFGu7vwU6a92cw.png)
+
+New build output structure will look like following:
+
+```sh
+|- <package-name>/
+|-  lib/  # module: ESNext
+|-  |- *.js
+|-  |- *.map.js
+|-  lib-commonjs/ # module: CommonJS
+|-  |- *.js
+|-  |- *.map.js
+|-  index.d.ts  # rolluped declaration file
+|-  package.json
+```
+
+**Problem 2 solution:**
+
+- run `declaration` generation only once during build.
+
 #### Pros
 
-- TBA
+- explicit public API contract
+- opportunity to leverage TypeScript language constructs beyond consumers support
+- less files shipped over network
+- bit faster build process
 
 #### Cons
 
-- TBA
+- additional configuration to maintain within our legacy tool chain
 
 ### 3. shipping only rollup-ed(bundled) implementation files
```

Comment:
This solution is what's mentioned as an approach for problem 1 of section 2, right?

![image](https://user-images.githubusercontent.com/39736248/163117077-35da25e3-4d06-4d96-9e23-ecf3be7266a0.png)

Reply:
no, there is nothing regarding implementation files in this PR/RFC